air/src/execution_step/instructions/fail.rs

```diff
@@ -127,11 +128,21 @@ fn fail_with_error(exec_ctx: &mut ExecutionCtx<'_>) -> ExecutionResult<()> {
         error,
         tetraplet,
         provenance,
+        orig_catchable,
     } = exec_ctx.error_descriptor.error();
 
     check_error_object(error).map_err(CatchableError::InvalidErrorObjectError)?;
 
-    fail_with_error_object(exec_ctx, error.clone(), tetraplet.clone(), provenance.clone())
+    let result = match orig_catchable {
+        Some(orig_catchable) => {
+            let catchable_to_return = orig_catchable.clone();
+            let _ = fail_with_error_object(exec_ctx, error.clone(), tetraplet.clone(), provenance.clone());
+            Err(ExecutionError::Catchable(catchable_to_return.into()))
+        }
+        None => fail_with_error_object(exec_ctx, error.clone(), tetraplet.clone(), provenance.clone()),
+    };
+    exec_ctx.error_descriptor.disable_error_setting();
```

Comment:
Setting error should be disabled for all error variants (e.g., for an error from a scalar).